### DISCUSSION TITLE
Count Rollup Not Working

### DISCUSSION THREAD

**Discussion by jessbr** [jamessimone/apex-rollup#260] [Q&A]
Hi @jamessimone,
I have used Apex-Rollups for a number of First and Last functions and it's working great. I am also trying to use it for a count function to count the number of service files of a particular record type and add the count to a field on the contact record. Service Files are directly linked to contacts through a contacts lookup field. These were not calculating correcting when I tested so I added to my calc clause  Contact Id = the id of one contact. When I rolled up again it worked. But obviously, I can't have the CMDT that only rolls up based on one contact Id. Do you know why the count would only work when I add the contact Id to the where clause. I have attached an image of the CMDT without the where clause that includes the contact ID. Thanks!
<img width="723" alt="ApexRollup" src="https://user-images.githubusercontent.com/63937108/157106096-ed66b191-6ed9-4099-919f-a6b521cdb1d1.png">

**Comment by jamessimone** [comment]
@jessbr interesting, was the Calc Item Where Clause you were attempting to use something like `RecordType.Name = ‘name of the RT’`? Or how was it set up without the singular Id?

**Reply by jessbr** [reply]
HI James, the example I gave you just had record Id = the name of the record ID in question (intake service files) and excluded the singular ID for the contact. I ran this query in workbench and it worked but it doesn't seem to be running in the count rollup. Should I try RecordType.Name = 'Intake' instead?

**Reply by jamessimone** [reply]
Ah I think I see the issue - I think you need to remove the `Changed Fields On Calc Item` text. You only need to specify either the changed fields _or_ a calc item where clause; I need to put something in the README about this because `apex-rollup` doesn't play well with both. I would try that

**Reply by jessbr** [reply]
That didn't do it either. It's almost like it's not reading the record type in the where clause. I looked at my other rollups and even if they have a recordtype Id specified in the Where Clause, they also have other criteria that would limit the rollup correctly. I change the recordId to recordtype name in the Where Clause and it's still not rolling up correctly.

**Reply by jamessimone** [reply]
@jessbr is there any chance you can enable the `Is Rollup Logging Enabled` checkbox on Rollup Control and get a log sent over? I would be happy to take a look but it may be difficult to diagnose further as is.

**Reply by jessbr** [reply]
I enabled it. Is it just the debug log you want and is this the one you'd need:

54.0 APEX_CODE,DEBUG;APEX_PROFILING,INFO;CALLOUT,INFO;DB,INFO;NBA,INFO;SYSTEM,DEBUG;VALIDATION,INFO;VISUALFORCE,INFO;WAVE,INFO;WORKFLOW,INFO
11:25:55.0 (960042)|USER_INFO|[EXTERNAL]|0055Y00000G9ipc|projects+dhrc@northpeak.com|(GMT-05:00) Eastern Standard Time (America/New_York)|GMT-05:00
11:25:55.0 (1142959)|EXECUTION_STARTED
11:25:55.0 (1209612)|CODE_UNIT_STARTED|[EXTERNAL]|01p5Y00000GkDJm|RollupFullBatchRecalculator
11:25:55.0 (699130439)|USER_DEBUG|[77]|DEBUG|Rollup v1.4.4: RollupFullBatchRecalculator finished
Class.RollupFullBatchRecalculator.finish: line 40, column 1

11:25:55.712 (712835603)|CUMULATIVE_LIMIT_USAGE
11:25:55.712 (712835603)|LIMIT_USAGE_FOR_NS|(default)|
  Number of SOQL queries: 0 out of 200
  Number of query rows: 0 out of 50000
  Number of SOSL queries: 0 out of 20
  Number of DML statements: 0 out of 150
  Number of Publish Immediate DML: 0 out of 150
  Number of DML rows: 0 out of 10000
  Maximum CPU time: 0 out of 60000
  Maximum heap size: 0 out of 12000000
  Number of callouts: 0 out of 0
  Number of Email Invocations: 0 out of 10
  Number of future calls: 0 out of 0
  Number of queueable jobs added to the queue: 0 out of 1
  Number of Mobile Apex push calls: 0 out of 10

11:25:55.712 (712835603)|CUMULATIVE_LIMIT_USAGE_END

11:25:55.0 (713046718)|CODE_UNIT_FINISHED|RollupFullBatchRecalculator
11:25:55.0 (713088213)|EXECUTION_FINISHED

**Reply by jamessimone** [reply]
@jessbr there should be either a log associated with a Queueable / Batch Apex chunk which will contain most of the relevant information. You can see in the one you posted that it's just the log for the `finish` method of the full recalc app: 

```
Rollup v1.4.4: RollupFullBatchRecalculator finished
Class.RollupFullBatchRecalculator.finish: line 40, column 1
```

**Comment by jessbr** [comment]
Attached. Let me know if this works for you
[apex-07L5Y00004kZcqYUAS.log](https://github.com/jamessimone/apex-rollup/files/8217009/apex-07L5Y00004kZcqYUAS.log)

**Reply by jamessimone** [reply]
Unfortunately that's not the correct log. If you open it yourself in the Salesforce Developer Console and toggle the `Debug Only` checkbox, if you _don't_ see a ton of information coming from Apex Rollup (in particular, lines like `lookup item before rolling up ...` and `lookup item after rolling up ...` then either it's not the correct file, or logging is not enabled, or both. To enable logging for Apex Rollup, head to the Manage section of Rollup Control custom metadata in setup, and toggle the box for `Is Rollup Logging Enabled?`

**Reply by jessbr** [reply]
@jamessimone  What is the name of the class I should be debugging in the developer console? And which portion of the code should I be debugging. Rollup logging is enabled

<img width="683" alt="rolluplogging" src="https://user-images.githubusercontent.com/63937108/157877201-cebeb2f5-496f-42bf-84f3-984660d1e921.png">

**Reply by jamessimone** [reply]
The default developer console debugging should be enough - typically when you have it open, debug logs with the appropriate logging level are automatically generated. You will probably be looking for one spawned by QueueableHandler (it should be quite large!)

**Reply by jessbr** [reply]
Hi @jamessimone I still don't know if I'm getting you what you need since the queueablehandler in the log item was small. But I've attached the larger debug log downloads related to rollups + the queuable handler debug log. The serialbatch apex debug logs have a CPU time limit error
[apex-07L5Y00004kZjurUAC.log](https://github.com/jamessimone/apex-rollup/files/8248278/apex-07L5Y00004kZjurUAC.log)
[apex-07L5Y00004kZkSGUA0.log](https://github.com/jamessimone/apex-rollup/files/8248266/apex-07L5Y00004kZkSGUA0.log)
[apex-07L5Y00004kZkQGUA0.log](https://github.com/jamessimone/apex-rollup/files/8248267/apex-07L5Y00004kZkQGUA0.log)
[apex-07L5Y00004kZjuJUAS.log](https://github.com/jamessimone/apex-rollup/files/8248273/apex-07L5Y00004kZjuJUAS.log)

**Reply by jamessimone** [reply]
@jessbr based on those logs (which were quite helpful, thank you), I would try updating the `Org Default` record for Rollup Control CMDT that ships with Apex Rollup. It seems to be consistently timing out during the full recalc process _while_ checking the CPU limits. In other words, right before it would automatically defer processing, it times out - the `Batch Chunk Size` property on the default Rollup Control record should be all you need to change. I would try lowering it from 2000 to 1000 and see if that fixes everything.

**Comment by jessbr** [comment]
Thanks James. I reduced the size to 1000 and am still getting CPU Time Limit exceeded on two SerialBatchApexRangeChunkHandler. Attached. Should I reduce the batch size further?
[apex-07L5Y00004kZmshUAC.log](https://github.com/jamessimone/apex-rollup/files/8255623/apex-07L5Y00004kZmshUAC.log)
[apex-07L5Y00004kZmsgUAC.log](https://github.com/jamessimone/apex-rollup/files/8255608/apex-07L5Y00004kZmsgUAC.log)
[apex-07L5Y00004kZmshUAC.log](https://github.com/jamessimone/apex-rollup/files/8255610/apex-07L5Y00004kZmshUAC.log)

**Reply by jamessimone** [reply]
I will take a look at the logs but yes I would definitely try that. Do you have parent items with thousands of children? I’ve had some people need to reduce the batch size to 200 or so to deal with things like that

**Comment by jessbr** [comment]
Hi James,
Sorry for the delay. I was ooo for a bit. I tried reducing the batch size to ten and it still didn't work. I added a new critieria that the application submission date could not be blank. ( WHERE RecordTypeId = '0125Y000000CK1IQAW' AND Application_Submitted_Date__c != null) The record Type Id represents an Intake Service File. I then turned on the full recalc flow which calls the rollup apex action to count all intake service files. I tried to cre
[apex-07L5Y00004ka07bUAA.log](https://github.com/jamessimone/apex-rollup/files/8337695/apex-07L5Y00004ka07bUAA.log)
ate a new intake service file and I got 'Attempt to de-reference a null object' error. I have another apexrollup that counts another type of service file and it's working well but there are far fewer in number.  Any advice is welcome. Log attached. Thanks
[apex-07L5Y00004ka07bUAA.log](https://github.com/jamessimone/apex-rollup/files/8337696/apex-07L5Y00004ka07bUAA.log)
d

**Reply by jamessimone** [reply]
I will take a look at the log, but I would try using `RecordTypeId = '0125Y000000CK1IQAW' AND Application_Submitted_Date__c != null` - you shouldn’t need parentheses or starting with the word `WHERE` — that’s supplied for you. On mobile at the moment but if I have a chance to look tomorrow I will!

**Reply by jessbr** [reply]
hi @jamessimone  to clarify, I didn't include the where statement in the CMT rollup. I just included it in my comment to you. See image of CMT
<img width="757" alt="rollup" src="https://user-images.githubusercontent.com/63937108/159913084-bbb45801-02f3-45b5-862b-40ead7752331.png">

**Reply by jamessimone** [reply]
I also have to apologize for the delay - still planning to look into this.

**Reply by jessbr** [reply]
Thanks @jamessimone much appreciated.

**Reply by jamessimone** [reply]
@jessbr it looks like the same log (ending in `7bUAA`) was attached twice to your last message - however, in reviewing it, I am wondering if you are using the correct Flow action? It looks like you're using `Full Recalc CMDT-driven Invocable` but the `Comma-separated list of API Names of the Rollup__mdt Records you want to run` property isn't filled out? That's what's causing the null pointer exception.

That particular flow action was a community addition to Apex Rollup; I can make it safer (or at least have it return a human-readable error for the situation you're running into), but again I'm wondering if you're using the right action to begin with since that property is the _only_ parameter necessary to fill out for that particular flow action.

**Reply by jessbr** [reply]
@jamessimone  Where do I set the comma-separate list of API names? In the CMDT rollup record or in the flow? It's curious because I'm using this type of flow for 3 other rollups that are working as intended and I haven't had to set anything differently.

**Reply by jamessimone** [reply]
@jessbr it would be within the flow, but there are two flow actions with similar names and I suspect here you were intending the use the other one (which I think your other flows are already using). Otherwise you would have recognized the property I’m referring to, because it’s the _only_ one on the flow action you currently have set up. I will update the label so that there’s a clearer delineation between the two in the next release
